### PR TITLE
add new API: paddle.linalg.cov

### PR DESCRIPTION
**Pull request by zhiboniu**
### PR types

New features
### PR changes

APIs
### Describe

add new API: paddle.linalg.cov

```
        import paddle
        import numpy as np

        x = np.random.rand(3,4)
        xt = paddle.to_tensor(x)
        paddle.linalg.cov(xt)

        '''
        Tensor(shape=[3, 3], dtype=float64, place=CUDAPlace(0), stop_gradient=True,
            [[0.07918842, 0.06127326, 0.01493049],
                [0.06127326, 0.06166256, 0.00302668],
                [0.01493049, 0.00302668, 0.01632146]])
        '''
```

image
image

### CODE REVIEW

File: python/paddle/tensor/linalg.py [hunk 1]
```diff
@@ -920,6 +920,121 @@ def dot(x, y, name=None):
...
+    op_type = 'cov'
+    if len(x.shape) > 2 or len(x.shape) < 1:
+        raise ValueError(
+            "Input(x) only support N-D (1<=N<=2) tensor in cov, but received "
+            "length of Input(input) is %s." % len(x.shape))
+    check_variable_and_dtype(x, 'dtype', ['float32', 'float64'], 'cov')
+    nx = x
+    if len(x.shape) == 1:
+        nx = x.reshape((1, -1))
+    if not rowvar and x.shape[0] != 1:
```
Comment (jeff41404) [inline python/paddle/tensor/linalg.py]: if rowvar==False and x.shape[0] == 1, we do not transpose, right?
why do we need "x.shape[0] != 1"

File: python/paddle/tensor/linalg.py [hunk 2]
```diff
@@ -920,6 +920,121 @@ def dot(x, y, name=None):
...
+        if aweights.shape[0] != observation_num:
+            raise ValueError(
+                "The number of Input(aweights) should equal to x's dim[1]: {}, but received "
+                "size of Input(aweights) is {}.".format(observation_num,
+                                                        aweights.shape[0]))
+        if aweights.min() < 0:
+            raise ValueError(
+                "The value of Input(aweights) cannot be negtive, but received "
+                "min of Input(aweights) is {}.".format(aweights.min()))
+        if fweights is not None:
```
Comment (jeff41404) [inline python/paddle/tensor/linalg.py]: if w is not None: better?

File: python/paddle/fluid/tests/unittests/test_cov.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+                np_arr = np.random.rand(*self.shape).astype(dtype)
+                tensor = paddle.to_tensor(np_arr, place=p)
+                cov = paddle.cov(tensor,
+                                 rowvar=True,
+                                 ddof=True,
+                                 fweights=None,
+                                 aweights=None)
+                np_cov = numpy_cov(
+                    np_arr, rowvar=True, ddof=1, fweights=None, aweights=None)
+                self.assertTrue(np.allclose(np_cov, cov.numpy(), rtol=3e-3))
```
Comment (jeff41404) [inline python/paddle/fluid/tests/unittests/test_cov.py]: is rtol=3e-3 too loose?

File: python/paddle/fluid/tests/unittests/test_cov.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+                np_arr = np.random.rand(*self.shape).astype(dtype)
+                tensor = paddle.to_tensor(np_arr, place=p)
+                cov = paddle.cov(tensor,
+                                 rowvar=False,
+                                 ddof=True,
+                                 fweights=None,
+                                 aweights=None)
+                np_cov = numpy_cov(
+                    np_arr, rowvar=False, ddof=1, fweights=None, aweights=None)
+                self.assertTrue(np.allclose(np_cov, cov.numpy(), rtol=3e-3))
```
Comment (jeff41404) [inline python/paddle/fluid/tests/unittests/test_cov.py]: is rtol=3e-3 too loose?

File: python/paddle/tensor/linalg.py (hunk 2 above)
Comment (zhiboniu) [reply python/paddle/tensor/linalg.py]: done

File: python/paddle/tensor/linalg.py (hunk 1 above)
Comment (zhiboniu) [reply python/paddle/tensor/linalg.py]: 对于1-d tensor按单个输入处理，不做转置。转置对大于1-d的输入生效

File: python/paddle/fluid/tests/unittests/test_cov.py (hunk 3 above)
Comment (zhiboniu) [reply python/paddle/fluid/tests/unittests/test_cov.py]: float32计算精度较差，float64精度非常高

File: python/paddle/fluid/tests/unittests/test_cov.py (hunk 4 above)
Comment (zhiboniu) [reply python/paddle/fluid/tests/unittests/test_cov.py]: float32计算精度较差，float64精度非常高

**Review by XiaoguangHu01 — DISMISSED** [review]
LG API

File: python/paddle/tensor/linalg.py [hunk 5]
```diff
@@ -920,6 +920,121 @@ def dot(x, y, name=None):
     return out
 
 
+def cov(x, rowvar=True, ddof=True, fweights=None, aweights=None, name=None):
+    """
+    Estimate the covariance matrix of the input variables, given data and weights.
+
+    A covariance matrix is a square matrix, indicate the covariance of each pair variables in the input matix.
```
Comment (TCChenlong) [inline python/paddle/tensor/linalg.py]: matix -> matrix

File: python/paddle/tensor/linalg.py [hunk 6]
```diff
@@ -920,6 +920,121 @@ def dot(x, y, name=None):
...
+    For example, for an N-dimensional samples X=[x1,x2,…xN]T, then the covariance matrix 
+    element Cij is the covariance of xi and xj. The element Cii is the variance of xi itself.
+
+    Parameters:
+        x(Tensor): A N-D(N<=2) Tensor containing multiple variables and observations. By default, each row of x represents a variable. Also see rowvar below.
+        rowvar(Bool, optional): If rowvar is True (default), then each row represents a variable, with observations in the columns. Default: True
+        ddof(Bool, optional): If ddof=1 will return the unbiased estimate, and ddof=0 will return the simple average. Default: None
```
Comment (TCChenlong) [inline python/paddle/tensor/linalg.py]: ddof=1  -> ddof=True
ddof=0 -> ddof=False
Default: None -> Default: True.

File: python/paddle/tensor/linalg.py [hunk 7]
```diff
@@ -920,6 +920,121 @@ def dot(x, y, name=None):
...
+
+    Returns:
+        Tensor: The covariance matrix Tensor of the variables.
+
+    Examples:
+
+    .. code-block:: python
+
+        import paddle
+        import numpy as np
```
Comment (TCChenlong) [inline python/paddle/tensor/linalg.py]: 不引入 np 直接用 paddle.rand

File: python/paddle/tensor/linalg.py (hunk 5 above)
Comment (zhiboniu) [reply python/paddle/tensor/linalg.py]: done

File: python/paddle/tensor/linalg.py (hunk 6 above)
Comment (zhiboniu) [reply python/paddle/tensor/linalg.py]: done

File: python/paddle/tensor/linalg.py (hunk 7 above)
Comment (zhiboniu) [reply python/paddle/tensor/linalg.py]: done

**Review by TCChenlong — APPROVED** [review]
LGTM

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Review by jeff41404 — APPROVED** [review]
LGTM

**Pull request merged**